### PR TITLE
Suppress errors when sending to a nonexistent receiver

### PR DESCRIPTION
**Pull request by DaveCTurner**
If the receiving statsd isn't running I think it'd be preferable to quietly drop the stats on the floor rather than report an error on each metric sent.

This PR changes the code to use `sendAllTo` instead of `sendAll` to achieve this end: as the socket is not `connect`ed it doesn't receive any destination-unreachable responses.

### CODE REVIEW

**Comment by tibbe** [comment]
I specifically added the existing logging to catch this case (to help me and other debug why things are not working). How about we add a counter to the loop and only log every N (e.g. 100) times this happens? Should reduce log spam but still be helpful.

**Comment by DaveCTurner** [comment]
Fair enough.

Stopping after a set number of messages is a little tricky, particularly with this implementation. For instance, when do you reset your error counter back to 0?

How about using `connect` if the `debug` flag is set in `StatsdOptions` and not otherwise?

**Comment by DaveCTurner** [comment]
Sorry, misread your message. Thought you said log the first 100 rather than every 100th error!

My question about changing the behaviour when `debug` is set still stands.

**Comment by DaveCTurner** [comment]
Like this, for instance?

File: System/Remote/Monitoring/Statsd.hs [hunk 1]
```diff
@@ -132,19 +142,19 @@ forkStatsd opts store = do
         "unsupported address: " ++ T.unpack (host opts)
     emptySample = M.empty
 
-loop :: Metrics.Store   -- ^ Metric store
-     -> Metrics.Sample  -- ^ Last sampled metrics
-     -> Socket.Socket   -- ^ Connected socket
-     -> StatsdOptions   -- ^ Options
+loop :: Metrics.Store            -- ^ Metric store
```
Comment (tibbe) [inline System/Remote/Monitoring/Statsd.hs]: If you add an `Int` parameter to this function and pass it to and return it from `flushSample` you can do something like

```haskell
Socket.sendAll socket msg `catch` \ (e :: IOException) -> do
    when failCount `mod` 100 == 0 $ do
        T.hPutStrLn stderr $ "ERROR: Couldn't send message: " <>
            T.pack (show e) <> "(happened " <> show failCount <> " times)"
    return $! failCount + 1
```

**Comment by DaveCTurner** [comment]
Yes, got that, but it wouldn't help here as you don't get these errors if you never call `connect` on the socket in the first place.

**Comment by tibbe** [comment]
Sorry, I haven't found the time to think about this hard enough. Could we make this work in a way that the user gets periodically notified if things aren't working without spamming them too much?

**Comment by DaveCTurner** [comment]
Sure. Your suggestion works for that. Again I would think different behaviour between `debug = True` and `debug = False` makes sense: when `True` I reckon every message should be shown. I'll add that to the PR in the next week when I find the time

But I really would like a way to avoid calling `connect` on the socket so the failures are dropped in the kernel rather than coming all the way back to the app. Maybe another config option if not a function of the `debug` setting?

**Comment by Gabriel439** [comment]
We ran into exactly this issue at work and we would like to fix it.  I can pick up and continue where this pull request left off if nobody objects

**Comment by tibbe** [comment]
Fine with me.

**Comment by joneshf** [comment]
What's the status here? We also ran into this and would prefer not to see the output.

**Comment by DaveCTurner** [comment]
Sorry, I'm unlikely to get to this in the near future. @Gabriel439 did you get anywhere?

**Comment by Gabriel439** [comment]
I also haven't gotten to this yet but I'm still willing to work on this.  I had just forgotten about this

**Comment by joneshf** [comment]
What is it that needs to be done?

**Comment by Gabriel439** [comment]
@joneshf: I'm reviewing the code right now.  My understanding is that the intent of the PR in its current form is:

* When `debug = True`, preserve the original behavior and log every failed attempt
* When `debug = False`, don't log failed attempts

The part I'm still wrapping my head around is how `sendAllTo` is supposed to silence failed attempts.  I'm not still not clear on that

Along the way @tibbe asked to log every 100 failures instead of logging every time, but I believe that is orthogonal to this pull request.  Whether to log at all is a separate issue from how frequently we log when we do choose to log them.  As far as I can tell, all three of @DaveCTurner, @joneshf and I want the errors silenced completely, not reduced in frequency.

So my approach to this is to first verify whether or not `sendAllTo` is the correct and idiomatic way to silence these errors and if that's the case then I will just resolve merge conflicts and push to merge this as is.  If that's not the idiomatic solution, then I will update the PR accordingly

**Comment by DaveCTurner** [comment]
I'm struggling to find an authoritative reference for this, but here's the explanation. It's not about `send()` vs `sendTo()`, it's about whether the socket is `connect()`ed or not. If it's connected then undeliverable-message errors are reported as errors to the application, which end up as exceptions by the time they get to this code. If it's not connected then undeliverable-message errors are dropped in the kernel before they ever make it to userspace, so the application sees nothing.

Following on from this, if the socket is not `connect()`ed then you have to supply the destination address with each datagram, hence the need for `sendTo()` rather than `send()`.

**Comment by Gabriel439** [comment]
@tibbe: Could you merge this in its present state now that the merge conflicts have been resolved?

**Pull request merged**